### PR TITLE
WIP: feature(pu): add Go env and AlphaZero league training

### PR DESCRIPTION
**Pull request by puyuan1996**
- add go_env, related unittest
- add go mcts bot and alphazero/muzero config
- add league version of alphazero
- add ctree version of alphazero

### CODE REVIEW

File: lzero/mcts/ctree/ctree_alphazero/mcts_alphazero.cpp [hunk 1]
```diff
@@ -0,0 +1,334 @@
...
+    void _simulate(Node* node, py::object simulate_env, py::object policy_forward_fn) {
+        // std::cout << "position21 " << std::endl;
+    
+        while (!node->is_leaf()) {
+            int action;
+            std::tie(action, node) = _select_child(node, simulate_env);
+            if (action == -1) {
+                break;
+            }
+            simulate_env.attr("step")(action);
```
Comment (jayyoung0802) [inline lzero/mcts/ctree/ctree_alphazero/mcts_alphazero.cpp:225]: simulate_env 执行step后的board可以打印出来检查是否正确

File: lzero/mcts/ctree/ctree_alphazero/mcts_alphazero.cpp [hunk 2]
```diff
@@ -0,0 +1,334 @@
...
+
+        return std::make_pair(action, action_probs);
+    }
+
+    void _simulate(Node* node, py::object simulate_env, py::object policy_forward_fn) {
+        // std::cout << "position21 " << std::endl;
+    
+        while (!node->is_leaf()) {
+            int action;
+            std::tie(action, node) = _select_child(node, simulate_env);
```
Comment (jayyoung0802) [inline lzero/mcts/ctree/ctree_alphazero/mcts_alphazero.cpp:221]: 打印legal action查看每次select_child后是否输出正常。

File: lzero/mcts/ctree/ctree_alphazero/mcts_alphazero.cpp [hunk 3]
```diff
@@ -0,0 +1,334 @@
...
+            _simulate(root, simulate_env_copy, policy_forward_fn);
+            // std::cout << "position9 " << std::endl;
+            simulate_env_copy = py::none();
+            // std::cout << "position10 " << std::endl;
+        }
+
+        std::vector<std::pair<int, int>> action_visits;
+            // std::cout << "position11 " << std::endl;
+        for (int action = 0; action < simulate_env.attr("action_space").attr("n").cast<int>(); ++action) {
+            if (root->children.count(action)) {
```
Comment (jayyoung0802) [inline lzero/mcts/ctree/ctree_alphazero/mcts_alphazero.cpp:175]: 为啥是count还是std::find？

File: lzero/mcts/ctree/ctree_alphazero/node_alphazero.h [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+    int get_visit_count() {
+        return visit_count;
+    }
+
+public:
+    Node* parent;
+    float prior_p;
+    int visit_count;
+    float value_sum;
```
Comment (jayyoung0802) [inline lzero/mcts/ctree/ctree_alphazero/node_alphazero.h:54]: 统一用double?

File: lzero/mcts/ctree/ctree_alphazero/mcts_alphazero.cpp [hunk 5]
```diff
@@ -0,0 +1,334 @@
...
+            node->children[actions[i]]->prior_p = node->children[actions[i]]->prior_p * (1 - frac) + noise[i] * frac;
+        }
+    }
+
+    // 在MCTS类中定义_select_child和_expand_leaf_node函数
+    std::pair<int, Node*> _select_child(Node* node, py::object simulate_env) {
+    
+        int action = -1;
+        Node* child = nullptr;
+        double best_score = -9999999;
```
Comment (jayyoung0802) [inline lzero/mcts/ctree/ctree_alphazero/mcts_alphazero.cpp:68]: double 是浮点数 -9999999.0

File: lzero/mcts/ctree/ctree_alphazero/mcts_alphazero.cpp [hunk 6]
```diff
@@ -0,0 +1,334 @@
...
+        }
+        // std::cout << "position6 " << std::endl;
+
+        for (int n = 0; n < num_simulations; ++n) {
+            // std::cout << "position7 " << std::endl;
+            py::object simulate_env_copy = simulate_env.attr("clone")();
+            // std::cout << "position8 " << std::endl;
+            _simulate(root, simulate_env_copy, policy_forward_fn);
+            // std::cout << "position9 " << std::endl;
+            simulate_env_copy = py::none();
```
Comment (jayyoung0802) [inline lzero/mcts/ctree/ctree_alphazero/mcts_alphazero.cpp:168]: 这句话貌似多余了

File: lzero/mcts/ctree/ctree_alphazero/mcts_alphazero.cpp (hunk 2 above)
Comment (jayyoung0802) [inline lzero/mcts/ctree/ctree_alphazero/mcts_alphazero.cpp:221]: _select_child的返回值是指针，这儿为啥还需要tie

**Comment by puyuan1996** [comment]
We have a new polished [PR](https://github.com/opendilab/LightZero/pull/65).

**Pull request closed without merging**